### DISCUSSION TITLE
Using DisplayModels UserScaleFactor does not seem to affect MapScaleBar - what to do?

### DISCUSSION THREAD

**Discussion by eddiemuc** [mapsforge/mapsforge#1313] [Q&A] [answered]
Since few days we use DisplayModel's userscale factor to allow users to scale the map according to their individual needs. We call it on the TileRendererLayer:

```
TileRendererLayer rendererLayer = ...
rendererLayer.getDisplayModel().setUserScaleFactor(scaleFactor)
```

This works fine except for the MapScaleBar displayed in the lower left corner. While the map itself enlarges as expected, the shown distances are the same as before and thus shorter than reality.

Debugging into mapsforge code shows that after setting user scale factor as described above, the DisplayModel object used by the MapScaleBar has the changed user scale factor. But it seems this factor is not taken into account.

![image](https://user-images.githubusercontent.com/6909759/167290703-dcfd2073-544a-4c2e-bb61-cacd1b103710.png)

An example showing above behaviour: displayed map with user scale factor 1 (=no scale):

![image](https://user-images.githubusercontent.com/6909759/167290799-a402c818-92ff-4980-8c5d-38dd52417f8a.png)

Same map with user scale factor 2 (=doubled size):

![image](https://user-images.githubusercontent.com/6909759/167290756-c669d545-4249-4d1d-9a0e-fac20b39189b.png)

As can be seen, the Map Scale Bar (circled in red in lower left corner) stays the same size on screen and still shows 200m for this size, although the shown distance is now only half of that (100m) because map has scaled by factor 2.

We solved this by applying a custom DistanceUnitAdapter which scales the meterRatio according to used user scale factor. But is this the correct way to handle this, or is there another preferred way?

**Comment by eddiemuc** [comment]
Update: I just found out that it is enough to trigger a redraw of MapScaleBar. This can be done e.g. by re-applying same `DistanceUnitAdapter` again, e.g. with something like this:

`this.mapView.getMapScaleBar().setDistanceUnitAdapter(this.mapView.getMapScaleBar().getDistanceUnitAdapter());`

The important thing seems to be that the `MapScaleBar`-internal flag `redrawNeeded` is set to true. Is there maybe a cleaner way to achieve this?

**Comment by devemux86** [comment, marked as answer]
Changing core variables as `DisplayModel` scale factor
requires to reinitialize everything, map, layers / overlays.

If you want the scale bar to keep its size with different user scale factors,
(re)initialize it with the device scale factor (instead of both device + user):
```java
mapView.setMapScaleBar(new DefaultMapScaleBar(
    mapView.getModel().mapViewPosition,
    mapView.getModel().mapViewDimension,
    AndroidGraphicFactory.INSTANCE,
    mapView.getModel().displayModel,
    DisplayModel.getDeviceScaleFactor()));
```

Also `MapScaleBar` has `(re)drawScaleBar` methods.

---

[Cruiser](https://wiki.openstreetmap.org/wiki/Cruiser) asks to restart the app when the display scale is changed.
